File: packages/ts/src/codes.ts
import registry from "../../../shared/error-codes.json" with { type: "json" };
import type { Issue, Severity } from "./types.js";

interface CodeDef { severity: Severity; category: string; message: string; spec_ref?: string; }

// Compile-time guard: fails `tsc` if error-codes.json drifts from CodeDef.
const _registryShapeCheck: Record<string, CodeDef> = registry as Record<string, CodeDef>;

export const CODES: Record<string, CodeDef> = registry as Record<string, CodeDef>;

function fill(template: string, data: Record<string, unknown>): string {
  return template.replace(/\{(\w+)\}/g, (_, k) => {
    const v = data[k];
    return v !== null && v !== undefined ? String(v) : `{${k}}`;
  });
}

export function makeIssue(
  code: string, path: string,
  data: Record<string, unknown> = {}, frame?: string,
): Issue {
  const def = CODES[code];
  if (!def) throw new Error(`Unknown error code: ${code}`);
  return {
    code, severity: def.severity, message: fill(def.message, data), path,
    ...(frame !== undefined ? { frame } : {}),
    ...(def.spec_ref ? { spec_ref: def.spec_ref } : {}),
    ...(Object.keys(data).length ? { data } : {}),
  };
}

File: packages/py/ocf_validator/codes.py
import json
import re
from pathlib import Path

from .types import Issue

_REGISTRY = json.loads(
    (Path(__file__).resolve().parents[3] / "shared" / "error-codes.json").read_text()
)


def _fill(template: str, data: dict) -> str:
    def repl(m):
        k = m.group(1)
        v = data.get(k, None)
        return str(v) if v is not None else m.group(0)

    return re.sub(r"\{(\w+)\}", repl, template)


def make_issue(
    code: str,
    path: str,
    data: dict | None = None,
    frame: str | None = None,
) -> Issue:
    data = data or {}
    d = _REGISTRY[code]
    return Issue(
        code=code,
        severity=d["severity"],
        message=_fill(d["message"], data),
        path=path,
        frame=frame,
        spec_ref=d.get("spec_ref"),
        data=data,
    )
